File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}

        for ch in s:
            if ch in mapping.values():  # opening bracket
                stack.append(ch)
            elif ch in mapping:         # closing bracket
                if not stack or stack[-1] != mapping[ch]:
                    return False
                stack.pop()
            else:
                return False  # invalid character

        return not stack

File: 20-valid-parentheses/valid-parentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> mapping = new HashMap<>();
        mapping.put(')', '(');
        mapping.put('}', '{');
        mapping.put(']', '[');
        
        for (char ch : s.toCharArray()) {
            if (mapping.containsValue(ch)) { // opening bracket
                stack.push(ch);
            } else if (mapping.containsKey(ch)) { // closing bracket
                if (stack.isEmpty() || stack.peek() != mapping.get(ch)) {
                    return false;
                }
                stack.pop();
            } else {
                return false; // invalid character
            }
        }
        
        return stack.isEmpty();
    }
}

